--- a/kernel/nv-frontend.c	2014-12-01 23:26:56.000000000 +0000
+++ b/kernel/nv-frontend.c	2015-01-08 11:51:16.760546089 +0000
@@ -327,7 +327,11 @@ long nvidia_frontend_unlocked_ioctl(
     unsigned long i_arg
 )
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
+    return nvidia_frontend_ioctl(file->f_path.dentry->d_inode, file, cmd, i_arg);
+#else
     return nvidia_frontend_ioctl(file->f_dentry->d_inode, file, cmd, i_arg);
+#endif
 }
 
 long nvidia_frontend_compat_ioctl(
@@ -336,7 +340,11 @@ long nvidia_frontend_compat_ioctl(
     unsigned long i_arg
 )
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
+    return nvidia_frontend_ioctl(file->f_path.dentry->d_inode, file, cmd, i_arg);
+#else
     return nvidia_frontend_ioctl(file->f_dentry->d_inode, file, cmd, i_arg);
+#endif
 }
 
 int nvidia_frontend_mmap(
--- a/kernel/nv.c	2014-12-01 23:26:56.000000000 +0000
+++ b/kernel/nv.c	2015-01-08 11:53:05.122427804 +0000
@@ -1839,7 +1839,11 @@ nvidia_unlocked_ioctl(
     unsigned long i_arg
 )
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
+    return nvidia_ioctl(file->f_path.dentry->d_inode, file, cmd, i_arg);
+#else
     return nvidia_ioctl(file->f_dentry->d_inode, file, cmd, i_arg);
+#endif
 }
 
 /*
